[Work-in-Progress] Track progress through a sequence (#75)

Deep-Chill opened the pull request:
- Moved Sequences from #71 to a new file sequences.py
- Added API endpoints with CRUD operations for EnumeratedProgress objects to a new file progress.py
- Added new serializers for the API endpoints above
- Created front-end templates, Vue components, and views for tracking and updating a user's progress within a Sequence using 10 second pings.
- Other minor changes include slightly rearranging the URLs, as app/energy-utility/ was conflicting with "re_path(r'^', include('pages.urls')),", added _paginator from djaodjin-saas, and a custom filter tag to get items from a dictionary in the template

Deep-Chill committed:
Track progress through a sequence

Deep-Chill committed:
Minor changes

Deep-Chill committed:
Minor changes

smirolo commented on pages/api/progress.py:
```diff
@@ -0,0 +1,245 @@
+     **Tags**: progress
+    """
+    queryset = EnumeratedProgress.objects.all().order_by('rank')
+    http_method_names = ['get', 'post', 'delete', 'head', 'patch', 'options']
```
I am not sure this line is necessary based on the implementation of django-restframework `APIView.dispatch`.

smirolo commented on pages/api/progress.py:
```diff
@@ -0,0 +1,245 @@
+        """
+        sequence_slug = self.kwargs.get('sequence_slug')
+        username = self.kwargs.get('username')
+        queryset = EnumeratedProgress.objects.all()
```
Why not use `self.queryset` here?

smirolo commented on pages/api/progress.py:
```diff
@@ -0,0 +1,245 @@
+
+        obj = get_object_or_404(queryset,
+                                rank=rank,
+                                progress__sequence__slug=sequence_slug,
```
filter by `progress__sequence__slug` and `progress_user__username` haven't been done in `get_queryset` already?

smirolo commented on pages/api/progress.py:
```diff
@@ -0,0 +1,245 @@
+
+        .. code-block:: http
+
+            POST /api/progress/sequence1/user1/3/ping HTTP/1.1
```
from what I can understand, it seems the example should be `POST /api/progress/sequence1/user1/3` (no `/ping` suffix).

smirolo commented on pages/api/progress.py:
```diff
@@ -0,0 +1,245 @@
+        """
+        instance = self.get_object()
+        now = timezone.now()
+        time_increment = timedelta(seconds=15)
```
candidate for a `settings` variable.

smirolo commented on pages/api/progress.py:
```diff
@@ -0,0 +1,245 @@
+        if instance.last_ping_time:
+            time_since_last_ping = now - instance.last_ping_time
+            if time_since_last_ping >= time_increment:
+                instance.viewing_duration += time_increment
```
interesting. we increment the viewing_duration by `time_increment` and not `time_since_last_ping`.

smirolo commented on pages/api/sequences.py:
```diff
@@ -0,0 +1,298 @@
+
+        .. code-block:: http
+
+            POST /api/sequences/1/remove_element HTTP/1.1
```
Why not use `DELETE /api/sequences/{sequence}/1`?

smirolo commented on pages/urls/api/progress.py:
```diff
@@ -0,0 +1,50 @@
+from ... import settings
+
+urlpatterns = [
+    re_path(r'^(?P<sequence_slug>%s)$' % settings.SLUG_RE,
```
best to use `path('<slug:sequence>', ...`. I don't think a special `SLUG_RE` regex is necessary. Plus, it read better to use `{sequence}` in documentation than `{sequence_slug}`.

smirolo commented on pages/urls/views/__init__.py:
```diff
@@ -22,9 +22,10 @@
+from ...compat import include, re_path, path
 
 urlpatterns = [
+    path(r'app/sequence/', include('pages.urls.views.progress')),
```
`app/sequence/` instead of `app/sequences/`? Have to check for consistency with other views, like `^editables/`.

smirolo commented on pages/views/progress.py:
```diff
@@ -0,0 +1,110 @@
+LOGGER = logging.getLogger(__name__)
+
+class SequenceProgressView(TemplateView):
+    template_name = 'pages/sequence_progress.html'
```
Let's try to have the `template_name` match the URL path whenever possible. - i.e. `template_name = `pages/app/sequence/index.html` for example.

smirolo commented on pages/views/progress.py:
```diff
@@ -0,0 +1,110 @@
+            'user': user,
+            'sequence': sequence,
+            'elements': elements,
+            'element_titles': {e.rank: e.page_element.title for e in elements}
```
So we load the elements from the database here. Will we reload them again through the API endpoint?

smirolo reviewed (approved):
Looks functionally correct. could be improved to make documentation and maintenance easier.

Deep-Chill committed:
Code refactor

Refactored code to fix:
- [x]  Removed redundant http_method_names
- [x]  use self.queryset
- [x]  remove redundant refilter
- [x]  Use a direct POST url to ping enumeratedprogress instances to update their viewing duration
- [x]  Added a settings variable
- [x]  Use time elapsed since last ping duration instead of time_increment in order to update the ping with checks for inactivity.
- [x]  Use DELETE instead of POST to remove an element from a sequence
- [x]  Use slug:sequence instead of a Regex
- [x]  Change URL to app/sequences
- [x]  consistent naming for templates & urls
- [x]  reload elements: We first load the elements from the database and then again from the API endpoint because the API endpoint does not return all PageElements, only the ones the user has progress on. But now that I think about it, it might make more sense to just directly use a different existing API endpoint to get this data.
- [x]  Update documentation

smirolo commented on pages/api/sequences.py:
```diff
@@ -235,64 +234,49 @@ def add_element(self, request, slug=None):
+        if element_rank is not None:
+            try:
+                element = EnumeratedElements.objects.get(sequence=sequence, rank=element_rank)
+                element.delete()
```
Note to self: will have to test what happens when browsing through a sequence where ranks have gaps.

smirolo reviewed (approved):
Looks like all previous comments were taken into account. Thank you!
